modin/core/dataframe/pandas/exchange/dataframe_protocol/dataframe.py

```diff
@@ -0,0 +1,188 @@
+    df : PandasDataframe
+        A ``PandasDataframe`` object.
+    nan_as_null : bool, default:False
+        A keyword intended for the consumer to tell the producer
+        to overwrite null values in the data with ``NaN`` (or ``NaT``).
+        This currently has no effect; once support for nullable extension
+        dtypes is added, this value should be propagated to columns.
+    allow_copy : bool, default: True
+        A keyword that defines whether or not the library is allowed
+        to make a copy of the data. For example, copying data would be necessary
+        if a library supports strided buffers, given that this protocol
+        specifies contiguous buffers. Currently, if the flag is set to ``False``
+        and a copy is needed, a ``RuntimeError`` will be raised.
+    """
+
+    def __init__(
+        self,
+        df: PandasDataframe,
+        nan_as_null: bool = False,
+        allow_copy: bool = True,
+    ) -> None:
+        self._df = df
+        self._nan_as_null = nan_as_null
+        self._allow_copy = allow_copy
+
+    @property
+    def metadata(self) -> Dict[str, Any]:
+        return {"modin.index": self._df.index}
+
+    def num_columns(self) -> int:
+        return len(self._df.columns)
+
+    def num_rows(self) -> int:
+        return len(self._df.index)
+
+    def num_chunks(self) -> int:
+        return self._df._partitions.shape[0]
+
+    def column_names(self) -> Iterable[str]:
+        for col in self._df.columns:
+            yield col
+
+    def get_column(self, i: int) -> PandasProtocolColumn:
+        return PandasProtocolColumn(
+            self._df.mask(row_positions=None, col_positions=[i]),
+            allow_copy=self._allow_copy,
+        )
+
+    def get_column_by_name(self, name: str) -> PandasProtocolColumn:
+        return PandasProtocolColumn(
+            self._df.mask(row_positions=None, col_labels=[name]),
+            allow_copy=self._allow_copy,
+        )
+
+    def get_columns(self) -> Iterable[PandasProtocolColumn]:
+        for name in self._df.columns:
+            yield PandasProtocolColumn(
+                self._df.mask(row_positions=None, col_labels=[name]),
+                allow_copy=self._allow_copy,
+            )
+
+    def select_columns(self, indices: Sequence[int]) -> "PandasProtocolDataframe":
+        if not isinstance(indices, collections.Sequence):
+            raise ValueError("`indices` is not a sequence")
+
+        return PandasProtocolDataframe(
+            self._df.mask(row_positions=None, col_positions=indices),
+            allow_copy=self._allow_copy,
+        )
+
+    def select_columns_by_name(self, names: Sequence[str]) -> "PandasProtocolDataframe":
+        if not isinstance(names, collections.Sequence):
+            raise ValueError("`names` is not a sequence")
+
+        return PandasProtocolDataframe(
+            self._df.mask(row_positions=None, col_labels=names),
+            allow_copy=self._allow_copy,
+        )
+
+    def get_chunks(
```

Comment:
I wonder if we could somehow share the implementation of `get_chunks` between this and the column

Reply:
Yes, we could if we were to use this object as an internal member of the column. That is done in @dchigarev's PR. We can look at this optimization later.